io-engine/src/subsys/nvmf/subsystem.rs

```diff
@@ -165,6 +196,189 @@ fn make_sn<T: AsRef<[u8]>>(uuid: T) -> CString {
 }
 
 impl NvmfSubsystem {
+    pub fn extract_nexus_name(nqn: &str) -> Option<String> {
+        let vec: Vec<String> = nqn.split(':').map(|s| s.to_string()).collect();
+
+        if vec.len() != 2 {
+            return None;
+        }
+        Some(vec[1].clone())
+    }
+
+    fn reset_cb(success: bool, ctx: *mut c_void) {
+        if success {
+            info!(
+                "nvmf_event_handler: controller successfully reset {:?}",
+                success
+            );
+        } else {
+            error!("nvmf_event_handler: failed to reset controller ");
+        }
+        done_cb(ctx, success);
+    }
+
+    pub async fn reset_one_child(dev_name: String, ctx: *mut c_void) {
+        let ctrlr = match NVME_CONTROLLERS.lookup_by_name(&dev_name) {
+            Some(ctrlr) => ctrlr,
+            None => {
+                debug!(
+                    "device name  {:?} not a valid NVMe controller",
+                    dev_name
+                );
+                Self::reset_cb(false, ctx);
+                return;
+            }
+        };
+        debug!("nvmf_event_handler: nvme controller {:?} ", ctrlr);
+        let mut ctrlr = ctrlr.lock();
+        if ctrlr.reset(NvmfSubsystem::reset_cb, ctx, false).is_ok() {
+            debug!("nvmf_event_handler: Reset initiated for {:?} ", ctrlr);
+        } else {
+            error!(
+                "nvmf_event_handler: Failed to initiate reset for {:?} ",
+                ctrlr
+            );
+            Self::reset_cb(false, ctx);
+        }
+    }
+
+    pub async fn reset_nexus(nexus_name: String) {
+        let mut nexus = match nexus_lookup_mut(&nexus_name) {
+            Some(v) => v,
+            None => {
+                debug!("nqn {:?} is not having valid nexus", nexus_name);
+                return;
+            }
+        };
+
+        //Pause the Nexus
+        if nexus.as_mut().pause().await.is_err() {
+            error!("Failed to pause Nexus {:?}", nexus);
+            return;
+        }
+        info!("Paused the nexus - {:?} ", nexus_name);
```

Comment:
We postpone nexus reset if there is at least one active initiator: 

```
extern "C" fn nvmf_event_handler()
   ...
   if nexus.initiator_cnt() > 0 {
        debug!("Reset operation not supported - initiator(s) present");
        return;
   }
```

Should we perform this check the final time before resetting the nexus after the subsystem has been paused ?
There can be a small chance that the first initiator is added again after 'reset_nexus()'  has been scheduled, but before the subsystem is paused.

Reply:
Good thought! Probably, it is a narrow window, since by default initiator takes 10s to reconnect. Worst case it might time out again and reconnect. I have added this check for more future-looking, I am not clear on the impact of this reset in the multi-initiator scenarios where multiple initiators may be used for failover or load balancing.  If we allow multiple initiators in the future, I suspect that this code may result in all other initiators' timeout if one times out when we pause the IO until replicas are reconnected.